### PR TITLE
Fix debug race with ads 

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Fix https://github.com/istio/istio/issues/39302

### CODE REVIEW

File: pilot/pkg/xds/debug.go [hunk 1]
```diff
@@ -987,10 +987,30 @@ func (s *DiscoveryServer) forceDisconnect(w http.ResponseWriter, req *http.Reque
...
 
+func cloneProxy(proxy *model.Proxy) *model.Proxy {
+	if proxy == nil {
+		return nil
+	}
+
+	proxy.Lock()
+	defer proxy.Unlock()
+
+	out := &(*proxy)
```
Comment (dddddai) [inline pilot/pkg/xds/debug.go:998]: IIRC this doesn't actually perform a shallow copy, I guess it was optimized by the compiler

Comment (hzxuzhonghu) [reply pilot/pkg/xds/debug.go:998]: Where did you see that

Comment (dddddai) [reply pilot/pkg/xds/debug.go:998]: You can run a simple test:
```go
type a struct {
	num int
}

func main() {
	x := &a{1}
	y := &(*x)
	println(x.num)	// 1
	y.num = 2
	println(x.num)	// 2
}
```
but if you change it to
```go
func main() {
	x := &a{1}
	y := *x
	z := &y
	println(x.num) // 1
	z.num = 2
	println(x.num) // 1
}
```
the output will be different

Comment (hzxuzhonghu) [reply pilot/pkg/xds/debug.go]: Thanks for this important finding

File: pilot/pkg/xds/debug.go [hunk 2]
```diff
@@ -987,10 +988,31 @@ func (s *DiscoveryServer) forceDisconnect(w http.ResponseWriter, req *http.Reque
 	_, _ = w.Write([]byte("OK"))
 }
 
+func cloneProxy(proxy *model.Proxy) *model.Proxy {
```
Comment (howardjohn) [inline pilot/pkg/xds/debug.go:991]: Interesting! I didn't think of this..

The shallow copy may have other races, although I cannot think of any.

One alternative is to have another `debugChannel` along the `select {push, request}`. That way we serialize it; we can send on it a `[]string` of TypeUrls and then have a `debugResponseChannel` to send the responses back or something. Not sure if its better than this though, WDYT

**Review by howardjohn — APPROVED** [review]
LGTM, but we should probably add a test calling PushAll and all(or most) of the debug interfaces in parallel. We have had a lot of issues like this in the past as well

**Pull request merged**